src/Agent.Plugins/PipelineArtifact/FileContainerProvider.cs

```diff
@@ -0,0 +1,175 @@
+        {
+            // Example of resourceData: "#/7029766/artifacttool-alpine-x64-Debug"
+            var segments = resourceData.Split('/');
+
+            long containerId;
+            if (segments.Length == 3 && segments[0] == "#" && long.TryParse(segments[1], out containerId))
+            {
+                return new Tuple<long, string>(
+                    containerId,
+                    segments[2]
+                    );
+            }
+            else
+            {
+                var message = $"Resource data value '{resourceData}' was not expected.";
+                throw new ArgumentException(message, "resourceData");
+            }
+        }
+
+        private async Task DownloadFileContainerAsync(Guid projectId, BuildArtifact artifact, string rootPath, IEnumerable<string> minimatchPatterns, CancellationToken cancellationToken)
+        {
+            var containerIdAndRoot = ParseContainerId(artifact.Resource.Data);
+
+            var items = await containerClient.QueryContainerItemsAsync(
+                containerIdAndRoot.Item1,
+                projectId,
+                containerIdAndRoot.Item2
+                );
+
+            IEnumerable<Func<string, bool>> minimatcherFuncs = this.GetMinimatchFuncs(minimatchPatterns);
+            if(minimatcherFuncs !=null && minimatcherFuncs.Count() !=0)
+            {
+                items = this.GetFilteredItems(items, minimatcherFuncs);
+            }
+
+            // Group items by type because we want to create the folder structure first.
+            var groupedItems = from i in items
+                               group i by i.ItemType into g
+                               select g;
+
+            // Now create the folders.
+            var folderItems = groupedItems.SingleOrDefault(i => i.Key == ContainerItemType.Folder);
+            if (folderItems != null)
+            {
+                Parallel.ForEach(folderItems, (folder) =>
+                {
+                    var targetPath = ResolveTargetPath(rootPath, folder);
+                    Directory.CreateDirectory(targetPath);
+                });
+            }
+
+            // Then download the files.
+            var fileItems = groupedItems.SingleOrDefault(i => i.Key == ContainerItemType.File);
+            if (fileItems != null)
+            {
+                Parallel.ForEach(fileItems, (file) =>
+                {
+                    var targetPath = ResolveTargetPath(rootPath, file);
+                    DownloadFileFromContainerAsync(containerIdAndRoot, projectId, containerClient, file, targetPath, cancellationToken).Wait();
+                });
+            }
+        }
+
+        private async Task DownloadFileFromContainerAsync(Tuple<long, string> containerIdAndRoot, Guid scopeIdentifier, FileContainerHttpClient containerClient, FileContainerItem item, string targetPath, CancellationToken cancellationToken)
+        {
+            var stream = await containerClient.DownloadFileAsync(
```

Comment:
I don't see any error handling or retries in case of failures here? Is that done at the root level?